### ISSUE TITLE
Changes from the Trenches: Should We Automate Them?

### ISSUE THREAD

**Issue by hsanchez** [hsanchez/paper-memory#1]
## TL;DR

This paper introduces PythonChangerMinder, a tool for discovering code change patterns
in the revision histories of Python projects.

### Paper Link

[arxiv.org/pdf/2105.10157](https://arxiv.org/pdf/2105.10157.pdf)

### Author/Institution

- Yaroslav Golubev (Jetbrains Research)
- Jiawei Li (UC Irvine)
- Timofey Bryksin (Jetbrains Research)
- Viacheslav Bushev (Saint Petersburg State University)
- Iftekhar Ahmed (UC Irvine)

## Overview

This paper reminds the reader of the importance of code changes in software evolution analysis.
They assert that code changes can be unique, but they can often be repetitive and thus follow patterns.

The paper's motivation is that most of the existing work focusing on mining code changes only
target statically-typed languages like Java. Because of this, there is still a big knowledge gap
in the understanding of code changes in dynamically-typed languages like Python. This works aims
at reducing this gap by introducing the PythonChangeMiner tool.

PythonChangeMiner facilitates a scalable approach for mining code change patterns
from the revision history of projects written in Python. The tool uses a special
code representation called fine-grained program dependence graphs (fgPDGs).

Given this revision history as an input, the tool parses each commit in the input into graphs,
and then uses a recursive algorithm for detecting any repetitive patterns within the changes.
The authors validated PythonChangeMiner on a small dataset, investigated the discovered recurring patterns,
and then classified them into structurally and thematically patterns. This classification facilitates
the discrimination or characterization of domains under study.

## Contributions and Distinctions from Previous Works

**Contributions:** 

- The PythonChangeMiner tool.
- A study of code change patterns from 120 projects in 4 domains.
- A survey of the authors of the changes about the discovered change patterns.
- A list of change patterns with the highest potential to be turned into code change recommendations.

**Differences:** 

- No work before has been directed towards mining a large number of Python code changes and then studying them to assess their usefulness in code change recommendation.
-  Built upon CPatMiner's pattern recognition algorithm to support pattern evaluation of semantic code changes targeting Python code.

## Methods

- Uses PyDriller to abstract the history of a software project
- Representation of code changes as fgPDGs in the abstracted history
- Algorithm for generating fgPDGs for files
- Algorithm that uses GumTree to identify change nodes and use that to change graphs from pairs of fgPDGs.

## Results

- Discovered 7,481 patterns that altogether contain 37,623 samples of code.
- Smallest pattern has 3 samples, the largest pattern has 156 samples.
- Of these 7,481 patterns, 803 include samples from at least two different projects.
- Four evaluated domains: Web, Media, Data, and ML+DL. 
- 243 patterns include at least one sample from Web, 384 from Media, 399 from Data, and 365 from ML+DL.

## Comments

See Paper #2 for details on how Git repository data is extracted.

<details>
<summary><b>Show sketch of PythonChangeMiner (click to expand)</b></summary>

```python
# Input: Git revision history H
# Output: a characterization of discovered change patterns in H

fgPDFs = []
for commit in commits(revision_history)
      for file in files(commit):
            fgPDFs += [build_fgPDF(file)]

change_graphs = []
for fgPDF in fgPDFs:
      change_graphs += [build_change_graph(fgPDF, next_revision(fgPDF))]

change_graph_patterns = pattern_evaluation(change_graphs)
return characterization(change_graph_patterns)

```

</details>

<br>

**Comment by hsanchez** [comment]
## Sketch of the PythonChangeMiner Algorithm:
```python
PythonChangeMiner(revision_history)
    fgPDFs = []
    for commit in commits(revision_history)
          for file in files(commit):
                fgPDFs += [build_fgPDF(file)]

    change_graphs = []
    for  fgPDF in fgPDFs:
          change_graphs += [build_change_graph(fgPDF, next_revision(fgPDF))]

    change_graph_patterns = pattern_evaluation(change_graphs)
    return characterization(change_graph_patterns)
```
